### PR TITLE
NLB-4532: Add lua module

### PR DESCRIPTION
**Pull request by xynicole**
### Proposed changes

Add lua module to crossplane. Validation reference to https://github.com/openresty/lua-nginx-module/blob/1d2aeb309c592484808f9b7313bc798461a37591/src/ngx_http_lua_module.c and https://github.com/openresty/lua-nginx-module/tree/master?tab=readme-ov-file#directives

*_by_lua and *_by_lua_block directives need additional lexer, I will do it in other PR

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [ ] I have read the [`CONTRIBUTING`](https://github.com/nginxinc/nginx-go-crossplane/blob/main/CONTRIBUTING.md) document
- [ ] If applicable, I have added tests that prove my fix is effective or that my feature works
- [ ] If applicable, I have checked that any relevant tests pass after adding my changes
- [ ] I have updated any relevant documentation ([`README.md`](https://github.com/nginxinc/nginx-go-crossplane/blob/main/README.md))

### CODE REVIEW

File: analyze.go [hunk 1]
```diff
@@ -2440,4 +2440,166 @@ var directives = map[string][]uint{
...
+		ngxHTTPSrvConf | ngxHTTPSifConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConf2More,
+	},
+	"content_by_lua_file": {
+		ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
+	},
+	"server_rewrite_by_lua_file": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxConfTake1,
+	},
+	"rewrite_by_lua_file": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxHTTPLocConf | ngxHTTPLifConf | ngxConfTake1,
```
Comment (dakshinai) [inline analyze.go:2495]:
```suggestion
		ngxHTTPMainConf | ngxHTTPSrvConf | ngxConfTake1,
```
based on https://github.com/openresty/lua-nginx-module/blob/1d2aeb309c592484808f9b7313bc798461a37591/src/ngx_http_lua_module.c#L309

**Comment by dakshinai** [comment]
LGTM with reference to https://github.com/openresty/lua-nginx-module/blob/1d2aeb309c592484808f9b7313bc798461a37591/src/ngx_http_lua_module.c

**Comment by ornj** [comment]
I think this PR is incomplete and you should add some tests for covering the lexer processing some of the more interesting looking lua directives. Here's a rather terse example test that covers this config demonstrating at least one `*_by_lua` directive is handled just fine:

```nginx
http {
    server {
        set_by_lua $res ' return 32 + math.cos(32) ';
    }
}
```

```diff
diff --git a/lex_test.go b/lex_test.go
index 1b9504d..01303eb 100644
--- a/lex_test.go
+++ b/lex_test.go
@@ -226,6 +226,18 @@ var lexFixtures = []lexFixture{
 		{";", 15},
 		{"}", 16},
 	}},
+	{"lua-simple", []tokenLine{
+		{"http", 1},
+		{"{", 1},
+		{"server", 2},
+		{"{", 2},
+		{"set_by_lua", 3},
+		{"$res", 3},
+		{" return 32 + math.cos(32) ", 3},
+		{";", 3},
+		{"}", 4},
+		{"}", 5},
+	}},
 }
 
 func TestLex(t *testing.T) {

```

**Comment by ornj** [comment]
> which don't need additional rules in the lexer

That may be, but adding some additional tests that prove that fact would be nice. My example above demonstrates at least one `*_by_lua` directive that works without any changes to the lexer.

**Review by dakshinai — APPROVED** [review]
Approving this hoping that we enable subset of lua directives except for `by_lua_block` and `by_lua`

**Review by ornj — APPROVED** [review]

**Comment by ornj** [comment]
Approved, but I still would like it if there were some lexer tests. The analyzer tests are great, but some more coverage to prove the lexer works as well (even though we _know_ it should) would make this more complete.

**Comment by xynicole** [comment]
@ornj added a lexer test for Lua

**Comment by ornj** [comment]
@xynicole looks good!

**Pull request merged**